allennlp/models/language_model.py

```diff
@@ -0,0 +1,206 @@
+from typing import Optional, Dict
+
+from overrides import overrides
+import torch
+import torch.nn as nn
+
+from allennlp.common import Params
+from allennlp.data import Vocabulary
+from allennlp.models.model import Model
+from allennlp.modules import TextFieldEmbedder
+from allennlp.common.checks import check_dimensions_match
+from allennlp.nn import InitializerApplicator, RegularizerApplicator
+
+from allennlp.modules.lm_rnn import LMRNN
+from allennlp.modules.adaptive import AdaptiveSoftmax
+from allennlp.training.metrics.perplexity import Perplexity
+
+@Model.register("WordLM")
+class WordLM(Model):
+    """
+    This ``WordLM`` is a word level language model.
+
+    Parameters
+    ----------
+    vocab : ``Vocabulary``, required
+        A Vocabulary, required in order to compute sizes for input/output projections.
+    text_field_embedder : ``TextFieldEmbedder``, required
+        Used to embed the ``tokens`` ``TextField`` we get as input to the model.
+    rnn : ``LMRNN``, required
+        The Recurrent Neural Networks for language modeling.
+    softmax : ``AdaptiveSoftmax``
+        The adaptive softmax for predicting next words
+    proj: bool, optional (default=``True``)
+        whether add a linear projection before the softmax layer
+    relu: bool, optional (default=``True``)
+        whether add a ReLU non-linear transformation before the softmax layer
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 rnn: LMRNN,
+                 softmax: AdaptiveSoftmax,
+                 proj: bool=True,
+                 relu: bool=True,
+                 dropout: float = 0.0,
+                 batch_first: bool=True,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+
+        super().__init__(vocab, regularizer)
+
+        self.rnn = rnn
+        self.batch_first = batch_first
+        self.text_field_embedder = text_field_embedder
+
+        if dropout:
+            self.dropout = torch.nn.Dropout(dropout)
+            dropout_list = [self.dropout]
+        else:
+            self.dropout = None
+            dropout_list = []
+
+        if proj:
```

Comment:
We only need a projection if `self.rnn.get_output_dim() != softmax.get_input_dim()`, so it would be easier to remove `proj` as an option and just automatically add one if necessary.

Reply:
I think in some cases (e.g., for a densely connected RNNs), i would like to add a non-linear projection layer after RNN (even the output dim of rnn is the same with the input dim of softmax). Maybe we can leave ```proj``` as an option for more flexible?

Reply:
OK in that case, I agree we need to leave it.  How about making the default = False since that would be the most common use case with embedding dim == softmax dim.